Docs: New App plugin tutorial (#914)
grafana/plugin-tools: docusaurus/docs/create-a-plugin/extend-a-plugin/nested-plugins.md

import ScaffoldYarn from '@snippets/createplugin-scaffold.yarn.md';

Grafana app plugins can nest data sources, both frontend and backend, together with panel plugins so that you can provide a complete user experience.

## Before you begin

See the tutorial to [build an app plugin](../../tutorials/build-an-app-plugin) and scaffold an app plugin.

## Anatomy of nested plugins

Nested plugins are located inside the app plugin `src` folder. They generally follow the same structure of a plugin and have their own `plugin.json`, but they don't have their own `package.json` or `.config` folder.

Here's an example of a nested data source plugin:

```diff bash
./src
 ├── README.md
 ├── components
+├── nested-datasource
+│   ├── components
+│   │   ├── ConfigEditor.tsx
+│   │   └── QueryEditor.tsx
+│   ├── datasource.ts
+│   ├── img
+│   ├── module.ts
+│   ├── plugin.json
+│   └── types.ts
 ├── img
 │   └── logo.svg
 ├── module.ts
 └── plugin.json
```

:::note

Notice that the plugin `nested-datasource` doesn't have a `package.json` of its own. The name of the nested plugin folder is not important.

:::

## How to add a nested plugin to an app plugin

> 1. Generate the nested plugin:

Reviewer: ```suggestion
1. Create a new plugin that will become the nested plugin:
```

The phrasing "Generate the nested plugin" here seems confusing, because what you're actually doing is creating a brand new plugin that will later become a nested plugin.

   :::important

   Begin outside your app plugin's directory.

   :::

   Use the `create-plugin` tool to generate a new plugin:

   ```bash
   npx @grafana/create-plugin@latest
   ```

>    Carefully select the desired plugin type (data source or panel), provide a name, and use the same organization as your app plugin.

Reviewer 2: ```suggestion
  Select the desired plugin type (data source or panel), provide a name, and use the same organization as your app plugin.
```

1. Prepare the nested plugin:

   Rename the `src` folder of your newly generated plugin to a name that reflects its specific purpose (for example, `nested-datasource`).

> 1. Integrate into your app plugin:

Reviewer 2: Should reference what is possible with backend data sources / backend apps

Author: @sympatheticmoose I am not sure what you mean here. can you suggest something to add?

Reviewer 2: For example - if you want your app to nest a backend datasource, you would likely create the backend datasource separately for the scaffolded code and then need to copy more than just `src`. For example the zabbix plugin https://github.com/grafana/grafana-zabbix where the app doesn't claim to have a backend, but the nested DS does

Reviewer 2: The exact details I would speak to the team about what is and isnt possible, I just feel theres more missing here

Author: I added a section on top of this one. I don't think this specific step is the best place to explain about backend datasources.

   Copy the renamed `src` folder directly inside the `src` folder of your app plugin.
   You can safely disregard the other generated files (like `package.json`, `.config`, and so on). These files aren't needed within your app.

1. (Optional) Clean up your directory:

   You can now delete the entire directory of the initially generated nested plugin.

> With these steps, your app plugin now houses the source code for your nested data source or panel, ready for further development.

Reviewer: Are there further steps needed here to be able to use the new nested plugins, such as updating the `GF_LOG_FILTERS` and `GF_PLUGINS_ALLOW_LOADING_UNSIGNED_PLUGINS` environment variables in the `docker-compose.yaml` file?

We should probably also include instructions for updating the `provisioning/datasources/datasources.yml` file if the user was adding a new nested datasource plugin.

Author: There's no need to add the nested plugins to the allow unsigned. 

>We should probably also include instructions for updating the provisioning/datasources/datasources.yml file if the user was adding a new nested datasource plugin.

That is covered in other parts of the documentation. The process is the same.

Reviewer: > That is covered in other parts of the documentation. The process is the same.

I think a simple call out to the other docs page as a reminder would go a long way here, as a developer may not be aware that it's a step they need to take. This somewhat relates to the frustration I had when navigating the docs, there's an expectation of implied knowledge that the developer has already read other docs pages (or even knows they exist).

Author: that sounds good.

---
id: nested-plugins
title: Work with nested plugins
description: How to work with nested plugins, Grafana app plugins.
keywords:
  - grafana
  - plugins
  - plugin
  - advanced
  - apps
  - app plugins
  - nested
---

import ScaffoldNPM from '@snippets/createplugin-scaffold.npm.md';
import ScaffoldPNPM from '@snippets/createplugin-scaffold.pnpm.md';
import ScaffoldYarn from '@snippets/createplugin-scaffold.yarn.md';

Grafana app plugins can nest data sources, both frontend and backend, together with panel plugins so that you can provide a complete user experience.

## Before you begin

> See the tutorial to [build an app plugin](../../tutorials/build-an-app-plugin) and scaffold an app plugin.

Reviewer 3: ```suggestion
Scaffold an app plugin. For instructions, refer to the tutorial to [build an app plugin](../../tutorials/build-an-app-plugin).
```

## Anatomy of nested plugins

Nested plugins are located inside the app plugin `src` folder. They generally follow the same structure of a plugin and have their own `plugin.json`, but they don't have their own `package.json` or `.config` folder.

Here's an example of a nested data source plugin:

```diff bash
./src
 ├── README.md
 ├── components
+├── nested-datasource
+│   ├── components
+│   │   ├── ConfigEditor.tsx
+│   │   └── QueryEditor.tsx
+│   ├── datasource.ts
+│   ├── img
+│   ├── module.ts
+│   ├── plugin.json
+│   └── types.ts
 ├── img
 │   └── logo.svg
 ├── module.ts
 └── plugin.json
```

## When to use a nested plugin

> When you have a data source or panel plugin that you want to distribute along with your app plugin.

Reviewer 3: ```suggestion
When you have a data source or panel plugin that you want to distribute along with your app plugin,
```

> Each nested data source can have its own backend, independent of the app plugin's backend.

Reviewer 3: ```suggestion
each nested data source can have its own backend, independent of the app plugin's backend.
```

:::note

> Notice that the plugin `nested-datasource` doesn't have a `package.json` of its own. The name of the nested plugin folder is not important.

Reviewer 3: ```suggestion
Notice that the plugin `nested-datasource` doesn't have a `package.json` of its own. The name of the nested plugin folder isn't important.
```

:::

## How to add a nested plugin to an app plugin

1. Create a new plugin that will become the nested plugin:

   :::important

   Begin outside your app plugin's directory.

   :::

   Use the `create-plugin` tool to generate a new plugin:

   ```bash
   npx @grafana/create-plugin@latest
   ```

   Select the desired plugin type (data source or panel), provide a name, and use the same organization as your app plugin.

1. Prepare the nested plugin:

   Rename the `src` folder of your newly generated plugin to a name that reflects its specific purpose (for example, `nested-datasource`).

1. Integrate into your app plugin:

   Copy the renamed `src` folder directly inside the `src` folder of your app plugin.
   You can safely disregard the other generated files (like `package.json`, `.config`, and so on). These files aren't needed within your app.

> 1. (Optional) Add your datasource to your provisioned datasources:

Reviewer 3: ```suggestion
1. (Optional) Add your data source to your provisioned data sources:
```

>    If you're adding a nested datasource, remember to add it to the provisioned datasources YAML configuration file, see the documentation on [Provisioning Grafana](https://grafana.com/docs/grafana/latest/administration/provisioning/#data-sources) for more details.

Reviewer 3: ```suggestion
   If you're adding a nested data source, remember to add it to the provisioned datasources YAML configuration file. Refer to the documentation on [Provisioning Grafana](https://grafana.com/docs/grafana/latest/administration/provisioning/#data-sources) for more details.
```
